File: src/main/scala/CunyFirstClient.scala
import com.gargoylesoftware.htmlunit._
import com.gargoylesoftware.htmlunit.WebClient
import com.gargoylesoftware.htmlunit.html._
import java.net.URL
import com.gargoylesoftware.htmlunit.util.NameValuePair
import collection.JavaConversions.{asScalaBuffer, seqAsJavaList}

class CunyFirstClient extends WebClient {
  val url = new URL("https://hrsa.cunyfirst.cuny.edu/psc/cnyhcprd/GUEST/HRMS/c/COMMUNITY_ACCESS.CLASS_SEARCH.GBL")

  this.getOptions setCssEnabled false
  this.getOptions setJavaScriptEnabled true
  this.getOptions setThrowExceptionOnScriptError false
  this.getCookieManager setCookiesEnabled true
  this setIncorrectnessListener Silent

  val settings = new WebRequest(url, HttpMethod.POST)
  val page: HtmlPage = getPage(settings)

  //result pages don't have Selects
  def getSelect(id: String) = (page getElementById id).asInstanceOf[HtmlSelect]

  private lazy val school = getSelect("CLASS_SRCH_WRK2_INSTITUTION$42$")
  private lazy val term = getSelect("CLASS_SRCH_WRK2_STRM$45$")

  def dept = getSelect("SSR_CLSRCH_WRK_SUBJECT_SRCH$0")

  //set institution, term, and course number
  def setup(inst: String, semester: String) = {
    school.setSelectedAttribute(school.getOptionByText(inst), true)
    waitForBackgroundJavaScript(10000)
    term.setSelectedAttribute(term.getOptionByText(semester), true)
    waitForBackgroundJavaScript(10000)

    //has to be done after school and term are set
    //search for course numbers > 0
    val exact = getSelect("SSR_CLSRCH_WRK_SSR_EXACT_MATCH1$1")
    exact setSelectedAttribute(exact getOptionByValue "G", true)
    (page getElementById "SSR_CLSRCH_WRK_CATALOG_NBR$1").asInstanceOf[HtmlTextInput] setText "0"
  }

  def params = (asScalaBuffer(((page.getForms get 0) getWebRequest null).getRequestParameters) map (a => (a.getName, a.getValue)))
    .toMap.updated("ICAction", "CLASS_SRCH_WRK2_SSR_PB_CLASS_SRCH").updated("SSR_CLSRCH_WRK_SSR_OPEN_ONLY$chk$5", "N")

  def getResults(updatedSettings: java.util.List[NameValuePair]): HtmlPage = {
    settings.setRequestParameters(updatedSettings)
    getPage(settings)
  }
}

object Silent extends IncorrectnessListener {
  def notify(x1: String, x2: Object) = {}
}

File: src/main/java/CunyFirstClient.java
//package cunyfirst;

import com.gargoylesoftware.htmlunit.*;
import com.gargoylesoftware.htmlunit.WebClient;
import com.gargoylesoftware.htmlunit.html.*;
import com.gargoylesoftware.htmlunit.util.NameValuePair;
import java.util.List;
import java.util.HashMap;
import java.util.ArrayList;
import java.io.IOException;
import java.net.URL;

class CunyFirstClient extends WebClient {
    public CunyFirstClient() {
        getOptions().setCssEnabled(false);
        getOptions().setJavaScriptEnabled(true);
        getCookieManager().setCookiesEnabled(true);
        setIncorrectnessListener(new Silent());

        try {
            request = new WebRequest(
                    new URL("https://hrsa.cunyfirst.cuny.edu/psc/cnyhcprd/GUEST/HRMS/c/COMMUNITY_ACCESS.CLASS_SEARCH.GBL"),
                    HttpMethod.POST);
            searchPage = getPage(request);

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private WebRequest request = null;
    private HtmlPage searchPage = null;
    HashMap<String, String> parameters = null;

    public HtmlSelect getSelect(String id) {
        return (HtmlSelect) searchPage.getElementById(id);
    }


    //set institution, term, and course number
    public void setup(String school, String semester) {
        HtmlSelect inst = getSelect("CLASS_SRCH_WRK2_INSTITUTION$42$");
        inst.setSelectedAttribute(inst.getOptionByText(school), true);
        waitForBackgroundJavaScript(10000);

        HtmlSelect term = getSelect("CLASS_SRCH_WRK2_STRM$45$");
        term.setSelectedAttribute(term.getOptionByText(semester), true);
        waitForBackgroundJavaScript(10000);

        //has to be done after school and term are set
        //search for course numbers > 0
        HtmlSelect match = getSelect("SSR_CLSRCH_WRK_SSR_EXACT_MATCH1$1");
        match.setSelectedAttribute(match.getOptionByValue("G"), true);
        ((HtmlTextInput)searchPage.getElementById("SSR_CLSRCH_WRK_CATALOG_NBR$1")).setText("0");

        List<NameValuePair> list = searchPage.getForms().get(0).getWebRequest(null).getRequestParameters();

        parameters = new HashMap<>();
        for(NameValuePair p: list) {
            parameters.put(p.getName(), p.getValue());
        }

        parameters.put("ICAction", "CLASS_SRCH_WRK2_SSR_PB_CLASS_SRCH");
        parameters.put("SSR_CLSRCH_WRK_SSR_OPEN_ONLY$chk$5", "N");

    }

    public HtmlPage getResults(String dept) throws IOException {
        parameters.put("SSR_CLSRCH_WRK_SUBJECT_SRCH", dept);
        List<NameValuePair> newParams = new ArrayList<>();
        parameters.entrySet().stream().map(e -> new NameValuePair(e.getKey(), e.getValue())).forEach(newParams::add);

        request.setRequestParameters(newParams);
        return getPage(request);
    }
}

class Silent implements IncorrectnessListener {
        public void notify(String x, Object y) {}
}
